lib/output/switch.go

```diff
@@ -0,0 +1,378 @@
+by listing child outputs paired with conditions. In the following example, messages
+containing "foo" will be sent to both the ` + "`foo`" + ` and ` + "`baz`" + ` outputs. Messages containing
+"bar" will be sent to both the ` + "`bar`" + ` and ` + "`baz`" + ` outputs. Messages containing both "foo"
+and "bar" will be sent to all three outputs. And finally, messages that do not contain
+"foo" or "bar" will be sent to the ` + "`baz`" + ` output only.
+
+` + "``` yaml" + `
+output:
+  type: switch
+  switch:
+    outputs:
+    - output:
+        type: foo
+        foo:
+          foo_field_1: value1
+      condition:
+        type: text
+        text:
+          operator: contains
+          part: 0
+          arg: foo
+      fallthrough: true
+    - output:
+        type: bar
+        bar:
+          bar_field_1: value2
+          bar_field_2: value3
+      condition:
+        type: text
+        text:
+          operator: contains
+          part: 0
+          arg: bar
+      fallthrough: true
+    - output:
+        type: baz
+        baz:
+          baz_field_1: value4
+        processors:
+        - type: baz_processor
+  processors:
+  - type: some_processor
+` + "```" + `
+
+The switch output requires a minimum of two outputs. If no condition is defined for
+an output, it behaves like a static ` + "`true`" + ` condition. If ` + "`fallthrough`" + ` is set to
+` + "`true`" + `, the swith output will continue evaluating additional outputs after finding
+a match. If an output applies back pressure it will block all subsequent messages,
+and if an output fails to send a message, it will be retried continously until
+completion or service shut down. Messages that do not match any outputs will be
+dropped.
+		`,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SwitchConfig contains configuration fields for the Switch output type.
+type SwitchConfig struct {
+	Outputs []switchConfigOutputs `json:"outputs" yaml:"outputs"`
+}
+
+// NewSwitchConfig creates a new SwitchConfig with default values.
+func NewSwitchConfig() SwitchConfig {
+	return SwitchConfig{}
+}
+
+type switchConfigOutputs struct {
+	cond        types.Condition
+	Condition   *condition.Config `json:"condition" yaml:"condition"`
+	Fallthrough bool              `json:"fallthrough" yaml:"fallthrough"`
+	Output      Config            `json:"output" yaml:"output"`
+}
+
+//------------------------------------------------------------------------------
+
+// Switch is a broker that implements types.Consumer and broadcasts each message
+// out to an array of outputs.
+type Switch struct {
+	running int32
+
+	logger log.Modular
+	stats  metrics.Type
+
+	throt *throttle.Type
+
+	transactions <-chan types.Transaction
+
+	outputTsChans  []chan types.Transaction
+	outputResChans []chan types.Response
+	outputs        []types.Output
+	outputNs       []int
+	confs          []switchConfigOutputs
+	typeStr        string
```

Comment:
Doesn't look like this gets set anywhere.

Reply:
Actually it's also worth simplifying your metrics to only increment values for `broker.switch.foo`, the idea is that only a final output target increments the shorter `output.foo` paths, otherwise we're double counting messages that leave the service.